org.eclipse.xtext/src/org/eclipse/xtext/build/IncrementalBuilder.java

```diff
@@ -358,6 +370,76 @@ protected void generate(Resource resource, BuildRequest request, Source2Generate
 			}
 		}
 
+		/**
+		 * @since 2.28
+		 */
+		protected boolean isLoadedFromStorage(Resource resource) {
+			return resource instanceof StorageAwareResource && ((StorageAwareResource) resource).isLoadedFromStorage();
+		}
+
+		/**
+		 * @since 2.28
+		 */
+		protected void installSourceLevelURIs(List<URI> uris) {
+			Set<URI> sourceLevelURIs = new HashSet<>(uris.size());
+			ResourceSet resourceSet = request.getResourceSet();
+			for (URI uri : uris) {
+				if (isSource(uri)) {
+					sourceLevelURIs.add(uri);
+					unloadResource(uri, this::isLoadedFromStorage);
+				}
+			}
+			SourceLevelURIsAdapter.setSourceLevelUrisWithoutCopy(resourceSet, sourceLevelURIs);
+		}
+
+		/**
+		 * Return true if the given uri must be loaded from source.
+		 * 
+		 * @since 2.28
+		 */
+		protected boolean isSource(URI uri) {
+			IResourceServiceProvider provider = context.getResourceServiceProvider(uri);
+			return provider instanceof IResourceServiceProviderExtension
+					&& ((IResourceServiceProviderExtension) provider).isSource(uri);
+		}
+		
+		/**
+		 * Deletes the binary models for the given set of URIs. Might be customized if
+		 * IResourceStorage supports deletion
+		 *
+		 * @param toBeDeleted
+		 *            set of URIs, must not be {@code null}
+		 * 
+		 * @since 2.28
+		 */
+		protected void deleteBinaryResources(final List<URI> toBeDeleted) {
+		}
+
+		/**
+		 * Stores resource as a binary.
+		 *
+		 * @param resource
+		 *            resource to store, must not be {@code null}
+		 * @param fileSystemAccess
+		 *            the file system access, must not be {@code null}
+		 * 
+		 * @since 2.28
+		 */
+		protected void storeBinaryResource(final Resource resource, IFileSystemAccessExtension3 fileSystemAccess) {
+			if (resource instanceof StorageAwareResource) {
+				IResourceStorageFacade resourceStorageFacade = ((StorageAwareResource) resource)
+						.getResourceStorageFacade();
+				if (resourceStorageFacade != null) {
+					resourceStorageFacade.saveResource((StorageAwareResource) resource, fileSystemAccess);
+					SourceLevelURIsAdapter adapter = SourceLevelURIsAdapter
+							.findInstalledAdapter(resource.getResourceSet());
+					if (adapter != null) {
+						adapter.getSourceLevelURIs().remove(resource.getURI());
```

Comment:
As far as I can tell, adapter.getSourceLevelURIs returns an unmodifiable set. How can this work?

Reply:
Right, I did not see that. What do you think of lifting that restriction? I think it makes, so that the builder does not need to pass the original set around.
I have seen, that we have created our own adapter that does not have the restriction, exactly to adress this point.